### PR TITLE
Connor Ferguson TV network challenge

### PR DESCRIPTION
**Pull request by cpfergus1**

### CODE REVIEW

**Comment by KathleenScriver** [comment]
Good start here with you commits!

**Comment by KathleenScriver** [comment]
I would like to see commits not for adding classes, but for adding pieces of functionality. So after a great test is written, commit. Then when the method is written that makes that test pass, commit.

File: lib/character.rb [hunk 1]
```diff
@@ -0,0 +1,11 @@
+class Character
+
+attr_reader :name, :actor, :salary
+
+  def initialize(character)
+    character.each do |key, val|
+      instance_variable_set("@#{key}",val) unless val.nil?
+    end
```
Comment (KathleenScriver) [inline lib/character.rb:8]: This works... but how can we set up these instance variables in a much more concise and easy to read way? We can do `@name = character[:name]`.

File: lib/network.rb [hunk 2]
```diff
@@ -0,0 +1,34 @@
+require './lib/character'
+require './lib/show'
```
Comment (KathleenScriver) [inline lib/network.rb:2]: These requires only need to be in the test files.

File: lib/show.rb [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+  attr_reader :name, :creator,:characters
+
+  def initialize(name,creator, characters)
+    @name = name
+    @creator = creator
+    @characters = characters
+  end
+
+  def total_salary
+    @characters.sum {|ind| ind.salary}
```
Comment (KathleenScriver) [inline lib/show.rb:14]: Great job using the sum enumerable. I would like to see your block variables named something more descriptive of the data they are holding though. What are we enumerating through? An array of characters so each element represents one character.

File: lib/show.rb [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+  end
+
+  def highest_paid_actor
+    @characters.sort_by {|ind| -ind.salary}.first.actor
```
Comment (KathleenScriver) [inline lib/show.rb:18]: Great job using the sort_by enumerable. Can you think of another enumerable you could use to reduce some of the method chaining?

File: lib/network.rb [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+    @shows =[]
+  end
+
+  def add_show(show)
+    @shows << show
+  end
+
+  def main_characters
+    main = @shows.map do |show|
+            show.characters.map do |character|
```
Comment (KathleenScriver) [inline lib/network.rb:20]: Watch your indentation here

File: lib/network.rb [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+                if (character.name == character.name.upcase) && character.salary > 500_000
```
Comment (KathleenScriver) [inline lib/network.rb:21]: Great job using the conditions to test if the character is a main character.

File: lib/network.rb [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+    @name = name
+    @shows =[]
+  end
+
+  def add_show(show)
+    @shows << show
+  end
+
+  def main_characters
+    main = @shows.map do |show|
```
Comment (KathleenScriver) [inline lib/network.rb:19]: This works, but can you think of another enumerable to use that would help you simplify this logic?

File: lib/network.rb [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+                  character
+                end
+              end
+            end.flatten.reject {|ind| ind.nil?}
+  end
+
+  def actors_by_show
+    hash = {}
+    @shows.each do |show|
+      hash[show] = show.characters.map {|character| character.actor}
```
Comment (KathleenScriver) [inline lib/network.rb:31]: This works, great job figuring this out. However, do you already have a method in Show that will give a list of the actors? Think about reusing methods you already have written. :)

File: test/show_test.rb [hunk 9]
```diff
@@ -0,0 +1,18 @@
...
+require './lib/character'
+require './lib/show'
+class ShowTest <Minitest::Test
+
+  def test_show_has_attributes
+    kitt = Character.new({name: "KITT", actor: "William Daniels", salary: 1_000_000})
+    michael_knight = Character.new({name: "Michael Knight", actor: "David Hasselhoff", salary: 1_600_000})
+    knight_rider = Show.new("Knight Rider", "Glen Larson", [michael_knight, kitt])
+    assert_equal 'Knight Rider', knight_rider.name
+    assert_equal 'Glen Larson', knight_rider.creator
```
Comment (KathleenScriver) [inline test/show_test.rb:12]: All the attributes can be in one test, but then every method should have its own test.

File: test/network_test.rb [hunk 10]
```diff
@@ -0,0 +1,27 @@
...
+    knight_rider = Show.new("Knight Rider", "Glen Larson", [michael_knight, kitt])
+    leslie_knope = Character.new({name: "Leslie Knope", actor: "Amy Poehler", salary: 2_000_000})
+    ron_swanson = Character.new({name: "Ron Swanson", actor: "Nick Offerman", salary: 1_400_000})
+    parks_and_rec = Show.new("Parks and Recreation", "Michael Shur & Greg Daniels", [leslie_knope, ron_swanson])
+    nbc.add_show(knight_rider)
+    nbc.add_show(parks_and_rec)
+    assert_equal [knight_rider, parks_and_rec], nbc.shows
+    assert_equal [kitt], nbc.main_characters
+    expected = {knight_rider => ["David Hasselhoff", "William Daniels"], parks_and_rec =>["Amy Poehler", "Nick Offerman"]}
+    assert_equal expected, nbc.actors_by_show
```
Comment (KathleenScriver) [inline test/network_test.rb:24]: Each method needs its own test.

**Comment by cpfergus1** [comment]
> I would like to see commits not for adding classes, but for adding pieces of functionality. So after a great test is written, commit. Then when the method is written that makes that test pass, commit.

I find myself scrambling on time and end up rushing commit messages. I hope my project commit messages are more clear and inline with your direction. Realistically, how much time do you expect these challenges to take for an average student? I am not sure I am meeting that.

File: lib/network.rb (hunk 8 above)
Comment (KathleenScriver) [reply lib/network.rb:31]: Actually, what is the return value of this whole method? It's not quite what we wanted.

**Pull request closed without merging**